### PR TITLE
Upgrade search app to Elastic Search 5.4

### PR DESCRIPTION
**Pull request by ericholscher**
This upgrades elastic search, and fixes indexing so that it works on 5.4.

Also integrates the new 5.4 search code in ext. I'm not 100% convinced this shouldn't live in the .org codebase, but just following existing convention of where the docsearch code was. 

Test failures are unrelated to this change.

### CODE REVIEW

File: readthedocs/restapi/views/footer_views.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def get_version_compare_data(project, base_version=None):
     }
     if highest_version_obj:
         ret_val['url'] = highest_version_obj.get_absolute_url()
-        ret_val['slug'] = highest_version_obj.slug,
+        ret_val['slug'] = highest_version_obj.slug
```
Comment (humitos) [inline readthedocs/restapi/views/footer_views.py]: Removing this comma produces tests to fail... I had the same problem and I ended up writing

    (highest_version_obj.slug, )

File: readthedocs/settings/base.py [hunk 2]
```diff
@@ -116,6 +116,7 @@ def INSTALLED_APPS(self):  # noqa
         if donate:
             apps.append('django_countries')
             apps.append('readthedocsext.donate')
+            apps.append('readthedocsext.search')
```
Comment (humitos) [inline readthedocs/settings/base.py]: Why is this under `if donate:`?

File: readthedocs/urls.py [hunk 3]
```diff
@@ -86,9 +86,13 @@
 
 if 'readthedocsext.donate' in settings.INSTALLED_APPS:
     # Include donation URL's
-    groups.append([
-        url(r'^sustainability/', include('readthedocsext.donate.urls')),
-    ])
+    groups.insert(0,
```
Comment (humitos) [inline readthedocs/urls.py]: Why this change?

Seems to be the same, anyway.

File: readthedocs/urls.py [hunk 4]
```diff
@@ -86,9 +86,13 @@
...
+                  [url(r'^sustainability/', include('readthedocsext.donate.urls'))]
+                  )
+    for num, _url in enumerate(rtd_urls):
```
Comment (humitos) [inline readthedocs/urls.py]: Same question again. This is only if the donate app is installed, why? I don't find the relationship between donate app and search app.

**Comment by safwanrahman** [comment]
Is it possible to upgrade latest 6.x?

File: readthedocs/settings/base.py (hunk 2 above)
Comment (ericholscher) [reply readthedocs/settings/base.py]: I fixed that here, actually: https://github.com/rtfd/readthedocs.org/pull/3356/files#diff-5973fe59a4d52278d55f884d145bd5d7R11 :) Waiting on it to get merged.

File: readthedocs/restapi/views/footer_views.py (hunk 1 above)
Comment (safwanrahman) [reply readthedocs/restapi/views/footer_views.py]: I think the test should be fixed. Its expecting a tuple but it should get the slug in sting

Comment (ericholscher) [reply readthedocs/restapi/views/footer_views.py]: @safwanrahman agreed -- want to PR it to master, and I'll merge it in here?

**Comment by ericholscher** [comment]
@safwanrahman that breaks _a lot_ of what we do, so I want to keep things here until we have more time to do a full rewrite. This is an incremental step.

File: readthedocs/search/indexes.py [hunk 5]
```diff
@@ -220,25 +218,24 @@ def get_mapping(self):
...
                 'properties': {
-                    'id': {'type': 'long'},
-                    'name': {'type': 'string', 'analyzer': 'default_icu'},
-                    'description': {'type': 'string', 'analyzer': 'default_icu'},
-
-                    'slug': {'type': 'string', 'index': 'not_analyzed'},
-                    'lang': {'type': 'string', 'index': 'not_analyzed'},
-                    'tags': {'type': 'string', 'index': 'not_analyzed'},
-                    'privacy': {'type': 'string', 'index': 'not_analyzed'},
+                    'id': {'type': 'keyword'},
```
Comment (safwanrahman) [inline readthedocs/search/indexes.py]: Is not `id` is `integer`?

Comment (ericholscher) [reply readthedocs/search/indexes.py]: I believe the new ES doesn't let these vary types, or something. I was getting an error, and it doesn't really _need_ to be an ID, so we will just use it as a string like the others.

Comment (safwanrahman) [reply readthedocs/search/indexes.py]: @ericholscher Seems like they support it
https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html

File: readthedocs/settings/base.py (hunk 2 above)
Comment (agjohnson) [reply readthedocs/settings/base.py]: Merged #3356, this can be rebased

File: readthedocs/search/indexes.py (hunk 5 above)
Comment (agjohnson) [reply readthedocs/search/indexes.py]: It does seem strange that we're coercing to string here, but I'm not familiar enough with ES to say why token/string is better.

Comment (ericholscher) [reply readthedocs/search/indexes.py]: It's not that it's better, it just doesn't work otherwise. I don't fully understand it either.

Comment (ericholscher) [reply readthedocs/search/indexes.py]: It throws this error: `RequestError: TransportError(400, u'illegal_argument_exception', u'mapper [id] cannot be changed from type [long] to [keyword]')`

**Comment by xrmx** [comment]
I've rebased this on master here: https://github.com/italia/readthedocs.org/tree/upgrade-es-rebased

It fails for me against ES 5.6.8 with the following:
```
{u'index':
  {u'status': 400, u'_type': u'page', u'_index': u'readthedocs', u'error': {
    u'caused_by': {
      u'reason': u'For input string: "de3a71b6edf5001ece51f4ea16ca4758"', u'type': u'number_format_exception'
    },
    u'reason': u'failed to parse [id]', u'type': u'mapper_parsing_exception'
  },
```

**Comment by xrmx** [comment]
I've done some work here: https://github.com/italia/readthedocs.org/tree/upgrade-es-rebased

Now we are indexing again, updated the search regarding filtering and faceting (now called aggs) to what ES 5 expects. 

Still broken:
- search for single project returns nothing
- searching on the homepage returns the following error. Fun fact is that lang is already a keyword :unamused: 
```
TransportError(400, u'search_phase_execution_exception', u'Fielddata is disabled on text fields by default. Set fielddata=true on [lang] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead.')
```

**Comment by ericholscher** [comment]
Closing this, hoping to be replaced by #4183
